pkg/tbtc/chain.go

```diff
@@ -196,6 +198,59 @@ type HeartbeatRequestedEvent struct {
 	BlockNumber     uint64
 }
 
+// WalletCoordinatorChain defines the subset of the TBTC chain interface that
+// pertains specifically to the tBTC wallet coordination.
+type WalletCoordinatorChain interface {
+	// OnDepositSweepProposalSubmitted registers a callback that is invoked when
+	// an on-chain notification of the deposit sweep proposal submission is seen.
+	OnDepositSweepProposalSubmitted(
+		func(event *DepositSweepProposalSubmittedEvent),
+	) subscription.EventSubscription
+
+	// PastDepositSweepProposalSubmittedEvents fetches past deposit sweep
+	// proposal events according to the provided filter or unfiltered if the
+	// filter is nil. Returned events are sorted by the block number in the
+	// ascending order, i.e. the latest event is at the end of the slice.
+	PastDepositSweepProposalSubmittedEvents(
+		filter *DepositSweepProposalSubmittedEventFilter,
+	) ([]*DepositSweepProposalSubmittedEvent, error)
+
+	// GetWalletLock gets the current wallet lock for the given wallet.
+	// Returned values represent the expiration time and the cause of the lock.
+	// The expiration time can be UNIX timestamp 0 which means there is no lock
+	// on the wallet at the given moment.
+	GetWalletLock(
+		walletPublicKeyHash [20]byte,
+	) (time.Time, WalletAction, error)
+}
+
+// DepositSweepProposal represents a deposit sweep proposal submitted to the chain.
+type DepositSweepProposal struct {
+	WalletPubKeyHash [20]byte
+	DepositsKeys     []struct {
+		FundingTxHash      bitcoin.Hash
+		FundingOutputIndex uint32
+	}
+	SweepTxFee *big.Int
+}
+
+// DepositSweepProposalSubmittedEvent represents a deposit sweep proposal
+// submission event.
+type DepositSweepProposalSubmittedEvent struct {
+	Proposal          *DepositSweepProposal
+	ProposalSubmitter chain.Address
+	BlockNumber       uint64
+}
+
+// DepositSweepProposalSubmittedEventFilter is a component allowing to
+// filter DepositSweepProposalSubmittedEvent.
+type DepositSweepProposalSubmittedEventFilter struct {
+	StartBlock          uint64
+	EndBlock            *uint64
+	ProposalSubmitter   []chain.Address
```

Comment:
Why this is an array and not a single value?

Reply:
Contract binding allows passing an array as a filter to get proposals submitted by a given list of submitters. Here we are just externalizing this ability.

Reply:
Tested it and works as expected. Here is the explanation:

1. Suppose we use a slice `ProposalSubmitter` with two items, "PS1" and "PS2".
2. We call `PastDepositSweepProposalSubmittedEvents` method. Under the hood, we reach `FilterLogs` (https://github.com/ethereum/go-ethereum/blob/23bee16208718975f9b9e012949b8d4ee1223108/accounts/abi/bind/base.go#L407).
3. Within `FilterLogs`, a topic set is constructed (https://github.com/ethereum/go-ethereum/blob/23bee16208718975f9b9e012949b8d4ee1223108/accounts/abi/bind/base.go#L413). Basically, this is the event ID, followed with our `ProposalSubmitter` slice. This gives the `{{ID}, {PS1, PS2}}` pattern. Here is a playground demostrating it: https://go.dev/play/p/l1zW2gNAI0J
4. According to the section you mentioned (https://github.com/ethereum/go-ethereum/blob/23bee16208718975f9b9e012949b8d4ee1223108/interfaces.go#L172), the `{{ID}, {PS1, PS2}}` pattern translates to: "matches topic (ID) in first position AND (PS1 OR PS2) in second position". In simple words this is: "matches the DepositSweepProposalSubmittedEvent AND (PS1 OR PS2) as submitters"